bmi_sdk/bmi_sdk/bmi_variable.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import annotations
+
+from enum import Enum
+
+# Once python 3.8 is no longer supported, just get Annotated directly
+# from typing import Annotated, Any, Literal, Optional, Self, Union
+from typing import Any, Literal, Optional, Union
+
+import numpy as np
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    RootModel,
+    computed_field,
+    model_validator,
+)
+
+# Only needed on python 3.8
+from typing_extensions import Annotated, Self
+
+from .bmi_grid import Grid
+
+Integer = RootModel[Literal["int", "int16", "int32", "integer"]]
+Long = RootModel[Literal["int", "int64", "longlong"]]
+Float = RootModel[
+    Literal[
+        "float",
+        "float32",
+        "np.float32",
+        "numpy.float32",
+        "np.single",
+        "numpy.single",
+    ]
+]
+Double = RootModel[Literal["float", "float64", "np.float64", "numpy.float64"]]
+
+DataType = Union[Integer, Long, Float, Double]
+
+
+class GridLocation(str, Enum):
+    """Enumeration of variable grid location strings supported by BMI.
+
+    See https://bmi.readthedocs.io/en/stable/#get-var-location
+
+    """
+
+    node = "node"
+    edge = "edge"
+    face = "face"
+
+
+class BmiVariable(BaseModel):
```

Comment:
I've asked myself the same question, and I settled on using pydantic and validation preciely for cases as shown in the current tests/validator.  It is important when interacting with BMI that certain constraints are met, for example the grid meta data and the data values align -- otherwise callers of BMI functions which have to make the assumption that they are correct may get bad info/data from bugs in the implementation.  We can definitely discuss more of the tradeoffs and solution space between the different approaches if you want!

Reply:
> I've asked myself the same question, and I settled on using pydantic and validation preciely for cases as shown in the current tests/validator.

I figured that was the case hahah. Thanks for elaborating on your thought process! What i'm hearing is this is a defensive strategy to catch bugs / assert invariants in python code that uses this. You would rather have a little overhead to do validation and catch things in python. Is that more or less fair?